### DISCUSSION TITLE
Is there a canonical way to use Terragrunt's dynamic backend ability to get a state file per instance of a module?

### DISCUSSION THREAD

**Discussion by rayterrill** [gruntwork-io/knowledge-base#345] [Help] [answered]
We have a module that we call hundreds of times to build instances of a thing. Ideally, we want each instance to have its own state file because each one of these instances can have n resources.

[r:terragrunt](https://github.com/gruntwork-io/terragrunt)

**Comment by gcagle3** [comment]
Hi Ray! When you call the module, are you doing so from a specific folder structure or through CLI arguments? If using a specific folder structure, you might be able to leverage the [remote_state](https://terragrunt.gruntwork.io/docs/features/keep-your-remote-state-configuration-dry/) key to differentiate between deployments in the parent terragrunt.hcl file. 

First, assume you have a folder structure something like this:

```
└── deployment_root
    ├── module1
    │   └── terragrunt.hcl
    ├── module2
    │   └── terragrunt.hcl
    ├── module3
    │   └── terragrunt.hcl
    └── terragrunt.hcl
```

In the `deployment_root` folder, there would be a terragrunt.hcl file with a remote_state configuration similar to the below:

```
remote_state {
    backend = "s3"
    config = {
        bucket          = "S3-BUCKET-FOR-STATE-STORAGE"
        key             = "ROOT-FOLDER-NAME/${path_relative_to_include()}/terraform.tfstate"
        region          = "REGION"
        dynamodb_table  = "DYNAMODB-TABLE-FOR-TFLOCK"
    }
}
```
The key configuration with `${path_relative_to_include()}` will allow terragrunt to configure unique states for each module being deployed based on the local file path. From here, you would just need to configure each module's terragrunt.hcl file to reference the parent terragrunt.hcl file with a configuration similar to the below: 

```
terraform {
    source = "MODULE-SOURCE-LINK-AND-VER"
}

include "root" {
    path = find_in_parent_folders()
}
```

Once you have this in place, deploying `deployment_root/module1` will result in a unique state file being saved to the S3 bucket at `deployment_root/module1/terraform.tfstate` and so on. 

Do you think a solution like this would work for your application?

**Reply by rayterrill** [reply]
Is there any way to do this that isn't folder based/doesn't require each instance of the module to be in a separate folder?

What we have today is basically a single folder with a ton of config files, which get parsed into module instances. But since they all live in the same actual folder, the state is all in the same place. Some way to do a "virtual folder" is more what I was looking for - it seems like everything requires an actual folder to break apart the state.

**Comment by gcagle3** [comment, marked as answer]
Hm, I don't think I quite understand how this is working just yet. Would it be possible for you to share some basic examples of how your current structure is working with the config files you referenced?

**Reply by rayterrill** [reply]
Sure.

Basically we have a bunch of yaml files in a directory, representing github repo configurations (including a bunch of related things - labels, branch protection rules, etc).

We read in those files, then pass them as one giant chunk to a module, which iterates over each of them and configures them.
```

   
terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
    }
    github = {
      source  = "integrations/github"
      version = "~> 4"
    }
  }

  required_version = "1.0.9"
  backend "s3" {
    region         = "us-west-2"
    bucket         = "our-terraform-state"
    key            = ".github/repositories/developer/terraform.tfstate"
    dynamodb_table = "terraform-lock"
  }
}

locals {
  repository_data_parent_path = "${path.module}/../../../repositories"
  config = {
    repositories = [for path in fileset(local.repository_data_parent_path, var.repository_data_sub_path) : yamldecode(file("${local.repository_data_parent_path}/${path}"))]
  }
}

module "repositories" {
  providers = {
    github = github
  }
  source                   = "../../../modules/repositories"
  repository_data_sub_path = "developer/**/*.yaml"
}
```

Even if I break this up, though, to not pass all the repos in one big chunk to the module, and call the module for each repo (the refactor I was thinking about that started this thread) - I still don't quite see how I could get each of the repo configs to have their own state because it all seems path based.

What would be really cool is if we could pass some kind of arbitrary value into the backend configuration - or have it be based on an arbitrary variable passed into a module - something that allow us to build out state locations in our remote state backend (s3) without actually having that reflected in our github folder structure.

Let me know if that isn't still making sense and I can try to explain further.

**Reply by yorinasub17** [reply]
What you are looking for is unfortunately not possible with Terragrunt in the current form, because it is working against the operating model of Terragrunt and design philosophy.

The operating model of terragrunt is one terragrunt folder = one terraform root module = one terraform state file. Anything that tries to break that is working against terragrunt and will lead to a lot of pain and/or hacks.

For example, you might be able to achieve something like this using environment variables, and an external script that calls to Terragrunt. Consider the following Terragrunt config:

```hcl
remote_state {
  backend = "s3"
  config = {
    encrypt        = true
    bucket         = "example-multi-terragrunt-state-bucket"
    key            = "${get_env("STATE_FILE_PATH")}/terraform.tfstate"
    region         = local.aws_region
    dynamodb_table = "terraform-locks"
  }
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }
}

inputs = {
  config = yamldecode(file(get_env("REPOSITORY_CONFIG_FILE")))
}
```

Note how the state key is now dependent on the environment variable `STATE_FILE_PATH` instead of the relative path between includes. This bypasses the requirement of using folders to map state paths, allowing you to dynamically change the state between runs even in the same folder.

You could then write a script to do what you want that calls `terragrunt` in parallel with each of the repo files. E.g., here is a python script (in serial, not concurrently):

```py
import glob
import subprocess
import tempfile
for f in glob.glob("path/to/dir/with/config/files/*.yml"):
    # We have to create a tmpdir to run `terragrunt` calls, to avoid conflicts in `.terragrunt-cache` between concurrent runs (this messes up `.terraform`).
    with tempfile.TemporaryDirectory() as tmpdirname:
        shutil.copy("/path/to/terragrunt/module", tmpdirname)
        subprocess.run(["terragrunt", "apply"], cwd=tmpdirname, env={'STATE_FILE_PATH': get_state_file_path(f), 'REPOSITORY_CONFIG_FILE': f})
```

This would work, but then you would now be dependent on an external arbitrary build script and not `terragrunt` directly. This script is a key driver here, because it is necessary to work around the limitations of both `terragrunt` and `terraform` in how the working state is stored in the working directory of the tool. (e.g., running `terragrunt` concurrently in the same folder with different inputs would cause a lot of weirdness even if you set a different dynamodb lock, because the two runs will share the same `.terragrunt-cache` and `.terraform` folders which contain the state information/config).

I can come up with other ways as well (e.g., using a codegen tool like [cookiecutter](https://cookiecutter.readthedocs.io/en/1.7.2/)), but at the end of the day, you would be reliant on a new tool that is not `terragrunt`. This is what I mean by pain and hacks.

We have ideas to evolve Terragrunt in the future to make it easier to do what you want (see https://github.com/gruntwork-io/terragrunt/issues/759), but to be transparent, that is looking more and more like a completely different tool than `terragrunt` and thus a rewrite. It is not something we will be able to knock out in the short term.

---

So to summarize, unfortunately, what you are looking for is not possible with pure `terragrunt`, but might be possible with a few hacks and external tooling. We are also unlikely to implement a feature to support this use case any time soon, as it requires a complete change in Terragrunt's operating model, which would more or less be a complete rewrite (and thus a new tool).

**Reply by rayterrill** [reply]
Thanks for the detailed answer, @yorinasub17. I had a feeling this was the case, just wanted to confirm that we weren't misunderstanding. Very much appreciated.

**Comment by gcagle3** [comment]
Thank you for the explanation! I think I have a better understanding of the process now. Unfortunately I don't know of an immediate solution for this specific workflow, but let me do a little research (and check with colleagues) and I will get back to you!
